### PR TITLE
Jvenstad/do unified

### PR DESCRIPTION
**Pull request by jonmv**
@mpolden please review. 

This changes deployment orchestration to less eagerly upgrade failing versions -- comments in doc. 
The code which determines versions for deployment is also change to take the max of change's and currently deployed version, when both exist. This makes it possible to deploy with a change which, e.g.,  upgrades application but downgrades platform -- it will simply upgrade application, and leave platform alone.

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/Application.java [hunk 1]
```diff
@@ -175,21 +176,19 @@ public Version deployVersionFor(DeploymentJobs.JobType jobType, Controller contr
...
-                return deploymentJobs().lastSuccessfulApplicationVersionFor(jobType);
-        }
-
-        return change().application();
+        return Optional.ofNullable(change.application()
+                                         .filter(version -> ! (useCurrentVersion || current.filter(cv -> cv.compareTo(version) > 0).isPresent()))
+                                         .orElse(current
+                                                         .orElse(deploymentJobs().lastSuccessfulApplicationVersionFor(jobType)
+                                                                                 .orElse(change.application()
+                                                                                               .orElse(null)))));
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/Application.java]: I'll need to revisit this when it's more finalised ;)

**Comment by jonmv** [comment]
Hang on, double negation error inside! ;O

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java [hunk 2]
```diff
@@ -256,10 +254,25 @@ else if (completedAt.isPresent()) { // Step not complete, because some jobs rema
...
+        Optional<Instant> lastSuccess = application.deploymentJobs().successAt(application.change(), jobType);
+        if (lastSuccess.isPresent() || ! jobType.isProduction())
+            return lastSuccess;
+
+        Deployment deployment = application.deployments().get(jobType.zone(controller.system()).get());
+        return Optional.ofNullable(deployment).map(Deployment::at)
+                       .filter(ignored ->    ! (   application.change().upgrades(deployment.version())
+                                                || application.change().upgrades(deployment.applicationVersion()))
+                                          &&   (   application.change().downgrades(deployment.version())
+                                                || application.change().downgrades(deployment.applicationVersion())));
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: I think the excessive formatting makes this harder to read when the operators are "far away" from the references.

**Review by mpolden — APPROVED** [review]
LGTM

**Pull request merged**